src/modules/job-manager/job-manager.c

```diff
@@ -58,6 +58,7 @@ void disconnect_rpc (flux_t *h,
                      const flux_msg_t *msg,
                      void *arg)
 {
+    alloc_disconnect_rpc (h, mh, msg, arg);
```

Comment:
Just to be sure we are on the same page, a disconnect _request_ is sent any time a module or API client that has sent a request closes its handle.  Thus the job manager will receive one `job-manager.disconnect` request from the scheduler when it unloads (with sender matching scehduler).  It will get other `job-manager.disconnect` requests from other entities that may have sent it requests.

It's slightly confusing (to me) in the alloc case because the alloc request is sent from job-manager to scheduler.  But since the scheduler has previously sent a hello request to the job manager, a disconnect message willl be generated anyway.

One thing to ponder is whether we still need to handle any  "normal" error response to an alloc request as a "teardown" event.  Perhaps that is no longer needed, and it is the source of some confusion.  However we would need to update the RFC if that behavior changes.

Reply:
@grondo I see your point, w/ the scheduler disconnect, its specifically a module disconnecting, and cannot be anything else.  Vs. the general case where a client may have sent multiple requests to multiple services in the job-manager.

This of course assumes the scheduler would never send a request to any other service in the job-manager.  I think this is a safe assumption?

Reply:
I apologize, I think I was confused about how disconnect requests are generated. My main point was that if only one of the disconnect handlers applies to any given disconnect request, then why call them all? (However, I admit I could still be confused).

Reply:
> One thing to ponder is whether we still need to handle any "normal" error response to an alloc request as a "teardown" event.  Perhaps that is no longer needed, and it is the source of some confusion.

Mentally, I've always thought of disconnects as "garbage collection", correcting for "bad client behavior".    I would think of errors on alloc to be the scheduler "cleaning up properly"?

Or mentally, would the addition of a "goodbye" in the protocol be proper cleanup?

Reply:
Well, in general, the disconnect is one per client, so if there is the potential for a client to have state in multiple places that needs to be destroyed, you would need to call into all those places.  I would probably make the code work like that regardless of what we know about the scheduler's behavior patterns.

Reply:
> I apologize, I think I was confused about how disconnect requests are generated. My main point was that if only one of the disconnect handlers applies to any given disconnect request, then why call them all? (However, I admit I could still be confused).

Generally speaking it should call all of them, but my point was that the scheduler is sort of a special case.  I can't imagine a scheduler wanting to use any of the other job-manager services (including future services we don't yet have), therefore there's no need to call all the disconnect handlers.  But not sure if its a safe assumption that the scheduler would never use any other job-manager service.

Perhaps I can just add a comment on this in the code.  Its probably not worth coding up so much for this case.

Reply:
> Well, in general, the disconnect is one per client, so if there is the potential for a client to have state in multiple places that needs to be destroyed, you would need to call into all those places. I would probably make the code work like that regardless of what we know about the scheduler's behavior patterns.

Sounds good, i think we're all in agreement to leave this code as is then.

Reply:
Sorry my comment/question got you off track. A comment in the code would be useful I think. When I was reviewing it seemed obvious that a single client would only be using one of these requests at a time, but I guess that isn't an invariant.